app/assets/javascripts/discourse/lib/register-topic-footer-button.js.es6

```diff
@@ -0,0 +1,124 @@
+let _topicFooterButtons = [];
+
+export function registerTopicFooterButton(action) {
+  const defaultAction = {
+    // id of the action, required
+    id: null,
+
+    // icon displayed on the button
+    icon: null,
+
+    // local key path for title attribute
+    title: null,
+    translatedTitle: null,
+
+    // local key path for label
+    label: null,
+    translatedLabel: null,
+
+    // is this button disaplyed in the mobile dropdown or as an inline button ?
+    dropdown: false,
+
+    // class class appended to the button class attribute
+    class: null,
+
+    // computed properties which should force a button state refresh
+    // eg: ["topic.bookmarked", "topic.category_id"]
+    dependentKeys: [],
+
+    // should we display this button ?
+    displayed: true,
+
+    // is this button disabled ?
+    disabled: false,
+
+    // display order, higher comes first
+    priority: 0
+  };
+
+  const normalizedAction = Object.assign(defaultAction, action);
+
+  if (!normalizedAction.id) {
+    Ember.error(`Attempted to register a topic action: ${action} with no id.`);
+    return;
+  }
+
+  if (!normalizedAction.icon && !normalizedAction.title) {
+    Ember.error(
+      `Attempted to register a topic action: ${
+        action.id
+      } with no icon or title.`
+    );
+    return;
+  }
+
+  _topicFooterButtons.push(action);
+}
+
+export function getTopicFooterButtons() {
+  const dependentKeys = [].concat(
+    ..._topicFooterButtons.map(tfb => tfb.dependentKeys).filter(x => x)
```

Comment:
What is the `filter` filtering here? That's a bit confusing.

Reply:
That's some pattern to ensure we dont have any empty element

Reply:
I see, it's like `compact` in ruby. Makes sense.